---
title: "Structural Bioinformatics (Pt. 1)"
author: "Joshua Cheung"
date: "02/13/2022"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 1. Introduction ot the RCSB Protein Data Bank (PDB)

## PDB statistics

We download the CSV files from the PDB site and access it as follows.

```{r}
# Here we save the input data file into our Project directory.
PDB.data <- "Data Export Summary.csv"
# We assign the result of the above code to an object called PDB.df.
PDB.df <- read.csv(PDB.data, row.names=1)
PDB.df
```


> **Q1. What percentage of structures in the PDB are solved by X-Ray and Electron Microscopy.**

```{r}
# The percentage of structures in the PDB solved by X-ray is:
round((sum(PDB.df$X.ray)/sum(PDB.df$Total))*100, digits=2)
# The percentage of structures in the PDB solved by EM is:      
round((sum(PDB.df$EM)/sum(PDB.df$Total))*100, digits=2)
```

Thus, the percentage of structures in the PDB solved by X-Ray is 87.2% and the percentage of structures solved in the PDB solved by Electron Microscopy is 5.35%.

> **Q2. What proportion of structures in the PDB are protein?**

```{r}
# The proportion of structures in the PDB that are protein is:
round((sum(PDB.df[1,]) - PDB.df[1,"Total"])/sum(PDB.df$Total), digits=3)
```

Thus, the proportion of structures in the PDB that are protein is 0.873.

> **Q3. Type HIV in the PDB website search box on the home page and determine how many HIV-1 protease structures are in the current PDB?**

After typing HIV into the PDV website search box on the home page and filtering the results by HIV type 1, there were still approximately 837 results according to the PDB website.  However, it is difficult to say for certain whether all these results are relvant to what we are looking for.  Thus, we conclude that the answer to how many HIV-1 protease structures is not easy to determine based solely from a text search with keywords like "HIV" or "HIV-1 protease."  Using a sequence search would get a much more reliable set of results.

## The PDB format

We download the PDB file for the HIV-1 protease structure with the PDB identify 1HSG.  We can examine the contents of the downloaded file using the Terminal tab from within RStudio with the following command.

```{r}
# less ~/Downloads/1hsg.pdb  ## (use ‘q’ to quit)
# Un-comment the above code and enter it in the Terminal tab to examine the 
# contents of the downloaded file.
```

# 2. Visualizing the HIV-1 protease structure

## Using atom selections

> **Q4. Water molecules normally have 3 atoms. Why do we see just one atom per water molecule in this structure?**

If we want to select and display all water molecules as red spheres we can use the selection text "resname HOH" and use the VDW drawing method and name coloring method.  In this case we see "one atom" per water molecule in this structure because the VDW method draws the atoms as spheres with the radius for each atom set to the van der Waals radius.  In water, oxygen is much larger than hydrogen, so oxygen exhibits greater van der Waals forces as it has a larger electron cloud surface area.  By comparison, hydrogen exhibits much smaller van der Waals forces due to its much smaller electron cloud surface area.  Thus the van der Waal sphere for hydrogen is much smaller and is entirely dwarfed by oxygen's vaan der Waal sphere hence we only see one atom (oxygen) per water molecule.

> **Q5. There is a conserved water molecule in the binding site. Can you identify this water molecule? What residue number does this water molecule have (see note below)?**

Yes, we can identify this water molecule.  It's residue number is HOH308:0.

> **OPTIONAL: Generate and save a figure clearly showing the two distinct chains of HIV-protease along with the ligand. You might also consider showing the catalytic residues ASP 25 in each chain (we recommend Licorice for these side-chains). Upload this figure to Piazza for some extra credit.**

The figure showing distinct chains of the HIV-protease as well as the ligand and the catalytic residues has been posted to Piazza. 

> **DISCUSSION:Can you think of a way in which indinavir, or even larger ligands and substrates, could enter the binding site?**

Indinavir could enter the binding site by being analog or a derivative from the same class of molecule as the proteases natural ligand allowing indinavir to competitively inhibit the protease ligand.  Similarly, a larger ligand or substrate could  enter the binding site, by having the portion of the larger ligand of substrate that interacts with the binding site, be an analog of the proteases natural ligand.  In the case of larger ligands and substrate it is important to note that while the rest of the larger molecule that does not interact with the binding site, must not block any binding site interactions and must not cause steric hindrance.

## Sequence viewer extension [OPTIONAL]

> **Q6. As you have hopefully observed HIV protease is a homodimer (i.e. it is composed of two identical chains). With the aid of the graphic display and the sequence viewer extension can you identify secondary structure elements that are likely to only form in the dimer rather than the monomer?**

While looking through the sequence viewer, we see that there is a single instance of an amino acid residue involved in a secondary structure called an isolated bridge (labelled "B").  This indicates there is an isolated hydrogen bond connecting elsewhere, likely the other dimer.  Thus it is likely that the isolated bridge secondary structure is only forms in the dimer rather than the monomer.

# 3. Introduction to Bio3D in R

We load the Bio3D package by using the following code:

```{r}
# install.packages("bio3d") ## Un-comment to install if necessary.
library(bio3d)
```

## Reading PDB file data into R

We use the read.pdb() function to read a single PDB file with Bio3D.

```{r}
pdb <- read.pdb("1hsg")
```

We get a quick summary of the contents of the pdb object we just created using the following code:

```{r}
pdb
```

> **Q7. How many amino acid residues are there in this pdb object?**

There are 198 amino acid resiudes in this pdb object.

> **Q8. Name one of the two non-protein residues?**

One of the non-protein residues is MK1.

> **Q9. How many protein chains are in this structure?**

There are 2 protein chains in this structure.

We now note that the attribute of this pdb object are listed in the last couple lines.  To find the attributes of any such pdb object we can use the following code.

```{r}
attributes(pdb)
```

Additionally, we note that in order to access these individual attributes we can use the dollar-attribute name convention.  For instance is we wished to access the atom attribute, we can use pdb$atom.

```{r}
# We preview the first 6 rows only as pdb is a large object.
head(pdb$atom)
```

# 4. Comparative structure analysis of adenylate kinase

## Setup

We begin this section by first installing the packages necessary.

```{r}
# We install the following packages in the R console and not this document.
## Un-comment any of the following lines to install if necessary

# install.packages("bio3d") 
# install.packages("ggplot2")
# install.packages("ggrepel")
# install.packages("devtools")
# install.packages("BiocManager")

# BiocManager::install("msa")
# devtools::install_bitbucket("Grantlab/bio3d-view")
```

> **Q10. Which of the packages above is found only on BioConductor and not CRAN?**

The msa package is found only in BioConductor and not CRAN.

> **Q11. Which of the above packages is not found on BioConductor or CRAN?**

The bio3d.view package is not found on BioConductor or CRAN.

> **Q12. True or False? Functions from the devtools package can be used to install packages from GitHub and BitBucket?**

True.

## Search and retrive ADK structures

We use the function get.seq() in order to fetch the query sequence for chain A of the PDB ID 1AKE.  We use this as an input to the function plst.pdb().

```{r}
library(bio3d)
aa <- get.seq("1ake_A")
```

We now examine the contents of "aa".

```{r}
aa
```

> **Q13. How many amino acids are in this sequence, i.e. how long is this sequence?**

There are 214 amino acids in this sequence.

We can now use this sequence to query to BLAST search the PDB in order to find similar sequences and stuctures.

```{r}
# We perform a blast or hmmer search.
b <- blast.pdb(aa)
```

We use the function plot.blast() to facilitate the visualization and filtering of the blast results.

```{r}
# We plot a summary of search results.
hits <- plot(b)
```

Next we proceed with only the top scoring hits (in black).

```{r}
# We list out some of the "top hits".
# We examine the first 6 elements of hits$pdb.id.
head(hits$pdb.id)
```

We can now use function get.pdb() and pdbslit() to fetch and parse the identified structures.

```{r}
# We now download the related pdb files.
files <- get.pdb(hits$pdb.id, path="pdbs", split=TRUE, gzip=TRUE)
```

## Align and superpose structures

We will now use the pdbaln() function to align and also optionally superimpose the identified pdb structures.

```{r}
# We first align the related pdbs.
pdbs <- pdbaln(files, fit = TRUE, exefile="msa")
```

```{r}
# We create a vector containing PDB codes for the figure axis.
ids <- basename.pdb(pdbs$id)
# We now draw the  alignment.
plot(pdbs, labels=ids)
```

Thus this figure is a schematic representation of the alignment. The grey regions depict aligned residues, while white depict gap regions. The red bar at the top depict sequence conservation.

## Optional: Viewing our superposed structures

We view our superposed results with the new bio3d.view view() function:

```{r}
library(bio3d.view)
library(rgl)
view.pdbs(pdbs)
```

Note that the structure will not be shown on the PDF version of this document.  Nonetheless, we include the code so that the reader can see what was done.

## Optional: Annotate collected PDB structures.

We use the pdb.annotate() function as a convenient way of annotating the PDB files we have collected.  We use the function to annotate each structure to its source species.

```{r}
anno <- pdb.annotate(c("2mh3_A", "4f3l"), anno.terms = c("structureId",
                                                         "experimentalTechnique"
                                                         , "resolution","pfam", 
                                                         "source", "citation"))
unique(anno$source)
```

We can now view all available annotation data as follows.

```{r}
anno
```

## Principal component analysis

We can perform PCA on the structural ensemble stored in the pdbs object with the function pca() or pca.xyz().  We use pca() below.

```{r}
# We now perform PCA.
pc.xray <- pca(pdbs)
plot(pc.xray)
```

We can then use the rmsd() function to calculate all pairwise RMSD values of the structural ensemble, facilitating clustering analysis based on the pairwise strucural deviation.

```{r}
# We calculate the RMSD.
rd <- rmsd(pdbs)
# Now we perform structural-based clustering.
hc.rd <- hclust(dist(rd))
grps.rd <- cutree(hc.rd, k=3)
plot(pc.xray, 1:2, col="grey50", bg=grps.rd, pch=21, cex=1)
```

The plot above is obtained by projecting the individual structures onto two selected PCs. These projections display the inter-conformer relationship in terms of the conformational differences described by the selected PCs.  This is called a conformer plot and is a low-dimensional representation of the conformational variability within the ensemble of PDB structures. 

# 5. Optional further visualization

We visualize the major structural variations in the ensemble with the function mktrj().  This generates a trajectory PDB file by interpolating along a given PC.

```{r}
# We visualize first principal component.
pc1 <- mktrj(pc.xray, pc=1, file="pc_1.pdb")
```

We can not only view our results with VMD but also with the bio3d.view view() function as follows.  Note that once again the three dimensional view of our results will not be available to view on the PDF function.  Nevertheless we include the code used for the reader to see. 

```{r}
view.xyz(pc1)
```

We now edit the code to set the color to highlight the most variable regions as follows.

```{r}
view.xyz(pc1, col=vec2color( rmsf(pc1) ))
```

Finally, we can also plot our main PCA results using ggplot2.

```{r}
library(ggplot2)
library(ggrepel)
# We define a data frame.
df <- data.frame(PC1=pc.xray$z[,1], 
                 PC2=pc.xray$z[,2], 
                 col=as.factor(grps.rd),
                 ids=ids)
# We can now plot the results with ggplot2.
p <- ggplot(df) + 
  aes(PC1, PC2, col=col, label=ids) +
  geom_point(size=2) +
  geom_text_repel(max.overlaps = 20) +
  theme(legend.position = "none")
p
```

# 6. Normal mode analysis

We now use the function nma() for normal mode analysis.  This function can be used for both single structures and complete structure ensembles allowing characterization and comparison of flexibility profiles of related protein structures.

```{r}
# We find the NMA of all structures.
modes <- nma(pdbs)
```

Now we create a plot as follows:

```{r}
plot(modes, pdbs, col=grps.rd)
```

> **Q14. What do you note about this plot? Are the black and colored lines similar or different? Where do you think they differ most and why?**

In this this plot the black and colored line are clearly different at 2 specific regions.  There is clear divergence around the the 25-60 residue number range and from the 125-175 residue number range.  Aside from these two regions, the black and colored lines appear to be close to each other.  A reason for the divergence in the two specific region listed is that there are two major distinct conformational states for Adk. These two states differ by a collective low frequency displacement of two nucleotide-binding sites that display distinct flexibilities upon binding of nucleotides.  Hence we see two particular areas on the plot where the black and colored lines show clear diveregence.